demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/RewriteXUserFilter.java

```diff
@@ -0,0 +1,72 @@
+package com.linagora.apisix.plugin;
+
+import static com.linagora.apisix.plugin.TokenRevokedFilter.makeUnAuthorizedRequest;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Locale;
+import java.util.Optional;
+
+import org.apache.apisix.plugin.runner.HttpRequest;
+import org.apache.apisix.plugin.runner.HttpResponse;
+import org.apache.apisix.plugin.runner.filter.PluginFilter;
+import org.apache.apisix.plugin.runner.filter.PluginFilterChain;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@Component
+public class RewriteXUserFilter implements PluginFilter {
+    public static final String TMAIL_X_USER_HEADER_NAME = "X-User";
+    private static final String PRE_FILTER = "pre";
+    private static final String POST_FILTER = "post";
+    private final Logger logger = LoggerFactory.getLogger("RewriteXUserPlugin");
+    private final ObjectMapper json;
+
+    public RewriteXUserFilter() {
+        this.json = new ObjectMapper();
+    }
+
+    @Override
+    public String name() {
+        return "RewriteXUserFilter";
+    }
+
+    @Override
+    public void filter(HttpRequest request, HttpResponse response, PluginFilterChain chain) {
+        final String mode = request.getConfig(this);
+        logger.debug("Received a new request. Filter mode: {}", mode);
+        if (mode.equalsIgnoreCase(PRE_FILTER)) {
+            xUserHeaderConditionCheck(request, response);
```

Comment:
Hm why doing a pre check? I don't think we really care if present or not, we just make sure to override it after the oidc introspection check, no? I believe that's what krakend was doing. WDYT?

Reply:
> we just make sure to override it after the oidc introspection check

Correct, If `X-User` always was upserted after oidc checking. 
But I feel a bit risky here, not sure to give an exact case. 
(Maybe one day miss config somewhere when ops, or bug somewhere). Then the client can fake X-User to enter another user

This line will avoid it. 
// Apisix doesn't support the Headers allow feature. So I did it in a plugin

Reply:
If we see it as redundant 
We can un-use it by removing that in config.yaml 
```
      ext-plugin-pre-req:
        conf:
          - name: RewriteXUserFilter
            value: 'pre'
```

Reply:
IMO It would be more clear to use built-in APISIX plugin for removing the `X-User` header, and let the `RewriteXUserFilter` add the header after.

Reply:
IMO it would be clear to use APISIX built in plugin to remove the X-User header, and let RewriteXUserFilter plugin add the header after

Reply:
Alright, not against it :)